File: src/orchestrator.ts
import { readFile } from "node:fs/promises";
import { setTimeout as sleep } from "node:timers/promises";
import { CursorAgentError, type SDKAgent } from "@cursor/sdk";
import type { Run, SendOptions } from "@cursor/sdk";
import { connectAgent } from "./agent.js";
import { installConnectionGuard } from "./connection-guard.js";
import type { ResolvedConfig } from "./config.js";
import { printConfigSummary } from "./config.js";
import { loadAgentsDocument, relativeToRoot } from "./project.js";
import { buildPrompt } from "./prompts.js";
import {
  formatRunErrorSummary,
  logRunErrorDetails,
} from "./run-diagnostics.js";
import { consumeRunStream } from "./stream.js";
import {
  errorMessage,
  isAgentBusyError,
  isTransientError,
} from "./transient-errors.js";
import {
  overseerTriggerReason,
  shouldRunOverseer,
} from "./overseer.js";
import { findLatestFailedReview } from "./review-feedback.js";
import { agentSteeringForPhase } from "./spec-steering.js";
import { analyzeRoadmap } from "./spec-roadmap.js";
import { initialState, loadState, nextPhase, saveState } from "./state.js";
import type { CuriousConfig, CuriousState, CycleRecord, Phase } from "./types.js";

export interface OrchestratorOptions {
  verbose?: boolean;
  /** Run a single phase (develop, review, or sync) then exit. */
  once?: boolean;
  /**
   * Stop after this many full develop→review→sync rounds.
   * Omit for continuous mode (until Ctrl+C or config maxCycles).
   */
  cycles?: number;
  /** Stop when every ## Roadmap task checkbox is checked. */
  untilDone?: boolean;
}

const REMAINING_TASK_PREVIEW = 8;

function formatTaskIdList(ids: string[]): string {
  if (ids.length === 0) return "";
  if (ids.length <= REMAINING_TASK_PREVIEW) {
    return ids.join(", ");
  }
  const head = ids.slice(0, REMAINING_TASK_PREVIEW).join(", ");
  return `${head}, … (+${ids.length - REMAINING_TASK_PREVIEW} more)`;
}

export class Orchestrator {
  private agent?: SDKAgent;
  private stopping = false;
  private lastSummary?: string;
  /** True when we cancelled a run to recover from a dropped connection. */
  private recoveryCancel = false;

  constructor(
    private readonly config: ResolvedConfig,
    private readonly opts: OrchestratorOptions = {},
  ) {}

  requestStop(): void {
    this.stopping = true;
  }

  async run(): Promise<void> {
    printConfigSummary(this.config);

    let state = await loadState(this.config.projectRoot);
    state.running = true;
    await saveState(this.config.projectRoot, state);

    const shutdown = async () => {
      this.requestStop();
      if (state) {
        state.running = false;
        await saveState(this.config.projectRoot, state);
      }
    };

    process.once("SIGINT", shutdown);
    process.once("SIGTERM", shutdown);

    try {
      this.agent = await connectAgent(this.config, state.agentId);
      state.agentId = this.agent.agentId;
      await saveState(this.config.projectRoot, state);

      console.log(`[curious] agent ${this.agent.agentId} (${this.config.runtime})`);

      const cycleAtStart = state.cycle;
      const cyclesLimit = this.opts.cycles;
      let skipLoop = false;

      if (this.opts.once) {
        console.log("[curious] mode: single phase (--once)");
      } else if (this.opts.untilDone) {
        const initial = analyzeRoadmap(await readFile(this.config.specPath, "utf8"));
        this.logRoadmapStatus(initial);
        if (initial.complete) {
          console.log("[curious] roadmap already complete — nothing to do");
          skipLoop = true;
        } else {
          console.log(
            "[curious] mode: until done (stops when all ## Roadmap tasks are checked)",
          );
        }
      } else if (cyclesLimit !== undefined) {
        console.log(
          `[curious] mode: ${cyclesLimit} full cycle(s) (develop → review → sync), then stop`,
        );
      } else {
        console.log(
          "[curious] mode: continuous (Ctrl+C to stop; default is --until-done)",
        );
      }

      if (!skipLoop)
      do {
        state = await this.runPhase(state);
        if (this.stopping) break;

        if (this.opts.once) {
          this.printNextStepHint(state);
          break;
        }

        const completedCycles = state.cycle - cycleAtStart;
        const finishedRequestedCycles =
          cyclesLimit !== undefined &&
          state.phase === "develop" &&
          completedCycles >= cyclesLimit;

        if (finishedRequestedCycles) {
          console.log(
            `[curious] finished ${completedCycles} cycle(s) (developer → reviewer → sync)`,
          );
          this.printContinueHint(cyclesLimit);
          break;
        }

        if (this.opts.untilDone && (await this.roadmapJustCompleted(state))) {
          break;
        }

        if (cyclesLimit !== undefined && state.lastError) {
          console.log("[curious] cycle aborted due to phase error");
          break;
        }

        if (this.config.maxCycles > 0 && state.cycle >= this.config.maxCycles) {
          console.log(`[curious] reached maxCycles=${this.config.maxCycles}`);
          break;
        }

        if (state.phase === "develop" && state.cycle > cycleAtStart) {
          console.log(
            `[curious] next: cycle ${state.cycle} · develop (see spec Progress)`,
          );
          if (this.config.cycleDelayMs > 0) {
            console.log(`[curious] sleeping ${this.config.cycleDelayMs}ms`);
            await sleep(this.config.cycleDelayMs);
          }
        }
      } while (!this.stopping);
    } finally {
      process.removeListener("SIGINT", shutdown);
      process.removeListener("SIGTERM", shutdown);
      state = await loadState(this.config.projectRoot);
      state.running = false;
      await saveState(this.config.projectRoot, state);
      if (this.agent) {
        await this.agent[Symbol.asyncDispose]();
      }
    }
  }

  private async runPhase(state: CuriousState): Promise<CuriousState> {
    const phase = state.phase;
    const specBody = await readFile(this.config.specPath, "utf8");
    const agents = await loadAgentsDocument(
      this.config.projectRoot,
      this.config.cwd,
    );

    if (phase === "develop" && !agents) {
      console.warn(
        "[curious] warning: no AGENTS.md at project root or agent cwd — developer will lack style guidelines",
      );
    } else if (
      agents &&
      (phase === "develop" || phase === "review" || phase === "overseer")
    ) {
      console.log(`[curious] including AGENTS.md (${agents.relPath}) in prompt`);
    }

    if (phase === "overseer") {
      console.log(
        `[curious] overseer triggered: ${overseerTriggerReason(state, this.config)}`,
      );
    } else if (phase === "develop" && findLatestFailedReview(state.history)) {
      console.log(
        "[curious] injecting full failed review into develop prompt",
      );
    } else if (agentSteeringForPhase(specBody, phase)) {
      console.log(`[curious] injecting Agent steering into ${phase} prompt`);
    }

    const prompt = buildPrompt({
      phase,
      specPath: this.config.specPath,
      specRelPath: relativeToRoot(this.config.projectRoot, this.config.specPath),
      specBody,
      cycle: state.cycle,
      cwd: this.config.cwd,
      projectRoot: this.config.projectRoot,
      agents,
      lastSummary: this.lastSummary,
      history: state.history,
    });

    console.log(`\n[curious] ── cycle ${state.cycle} · ${phase} ──`);

    if (!this.agent) {
      throw new Error("Agent not connected");
    }

    const startedAt = new Date().toISOString();
    let runId = "";
    let status: CycleRecord["status"] = "error";
    let summary: string | undefined;

    try {
      const sendOptions: SendOptions = {
        onStep: this.opts.verbose
          ? ({ step }) => {
              console.log(`[curious] step ${step.type}`);
            }
          : undefined,
      };

      if (this.shouldForceLocalRun(state)) {
        sendOptions.local = { force: true };
        console.log("[curious] expiring any stuck local run before send");
        state.needsForceNextSend = false;
      }

      const run = await this.sendPrompt(prompt, sendOptions);
      runId = run.id;
      console.log(`[curious] run ${runId}`);
      console.log(
        "[curious] agent working (tool progress below; use --verbose for full text)…",
      );

      const statusStop = run.onDidChangeStatus((status) => {
        console.log(`[curious] run status → ${status}`);
      });

      this.recoveryCancel = false;
      const guard = installConnectionGuard(run, () => {
        this.recoveryCancel = true;
      });

      try {
        await consumeRunStream(run, {
          verbose: this.opts.verbose,
          progress: true,
          signal: guard.abort.signal,
        });

        const result = await run.wait();

        status = result.status;
        summary = result.result;
        this.lastSummary = summary;

        if (result.status === "error") {
          const errorReason = formatRunErrorSummary(result.result);
          state.lastError =
            errorReason ?? `Run ${runId} ended with error`;
          console.error(`[curious] run error: ${runId}`);
          await logRunErrorDetails(run);
        } else if (result.status === "cancelled") {
          if (this.recoveryCancel || guard.abort.signal.aborted) {
            throw guard.abort.signal.reason ?? new Error("connection lost");
          }
          console.log(`[curious] run cancelled: ${runId}`);
          this.stopping = true;
        } else {
          state.lastError = undefined;
          state.needsForceNextSend = false;
          console.log(
            `[curious] run finished in ${result.durationMs ?? "?"}ms`,
          );
          if (summary) {
            const preview =
              summary.length > 400 ? `${summary.slice(0, 400)}…` : summary;
            console.log(`[curious] result: ${preview}`);
          }
        }
      } finally {
        guard.dispose();
        statusStop();
        this.recoveryCancel = false;
      }
    } catch (err) {
      const message = errorMessage(err);

      if (isTransientError(err)) {
        state.lastError = message;
        state.needsForceNextSend = true;
        console.error(
          `[curious] transient error during ${phase} (will retry same phase in 10s): ${message}`,
        );
        await sleep(10_000);
      } else if (isAgentBusyError(err)) {
        state.lastError = message;
        state.needsForceNextSend = true;
        console.error(
          `[curious] agent busy (will retry ${phase} with force in 5s): ${message}`,
        );
        await sleep(5_000);
      } else if (err instanceof CursorAgentError) {
        state.lastError = message;
        console.error(
          `[curious] agent error: ${message} (retryable=${err.isRetryable})`,
        );
        if (err.isRetryable) {
          await sleep(10_000);
        } else {
          this.stopping = true;
        }
      } else {
        state.lastError = message;
        throw err;
      }
    }

    const record: CycleRecord = {
      cycle: state.cycle,
      phase,
      runId: runId || "unknown",
      status,
      startedAt,
      finishedAt: new Date().toISOString(),
      summary,
    };

    state.history.push(record);
    if (state.history.length > 100) {
      state.history = state.history.slice(-100);
    }
    state.lastRunId = runId || state.lastRunId;

    if (status === "finished") {
      if (phase === "sync") {
        state.cycle += 1;
        state.phase = shouldRunOverseer(state, this.config)
          ? "overseer"
          : nextPhase("sync");
      } else {
        state.phase = nextPhase(phase);
      }
    } else {
      console.log(`[curious] staying on phase "${phase}" until a run finishes successfully`);
    }

    await saveState(this.config.projectRoot, state);
    return state;
  }

  private shouldForceLocalRun(state: CuriousState): boolean {
    if (this.config.runtime !== "local") {
      return false;
    }
    if (state.needsForceNextSend || state.lastError) {
      return true;
    }
    const last = state.history.at(-1);
    return last !== undefined && last.status !== "finished";
  }

  private async sendPrompt(
    prompt: string,
    sendOptions: SendOptions,
  ): Promise<Run> {
    if (!this.agent) {
      throw new Error("Agent not connected");
    }

    try {
      return await this.agent.send(prompt, sendOptions);
    } catch (err) {
      if (
        this.config.runtime === "local" &&
        isAgentBusyError(err) &&
        !sendOptions.local?.force
      ) {
        console.log("[curious] agent busy — retrying send with force");
        return await this.agent.send(prompt, {
          ...sendOptions,
          local: { ...sendOptions.local, force: true },
        });
      }
      throw err;
    }
  }

  private logRoadmapStatus(status: ReturnType<typeof analyzeRoadmap>): void {
    if (status.totalTasks === 0) {
      console.log("[curious] roadmap: no T*/M* tasks found in ## Roadmap");
      return;
    }
    const remaining = formatTaskIdList(status.uncheckedTaskIds);
    console.log(
      `[curious] roadmap: ${status.checkedTasks}/${status.totalTasks} tasks done` +
        (remaining ? ` (remaining: ${remaining})` : ""),
    );
  }

  /** True after a successful sync when every roadmap task is checked off. */
  private async roadmapJustCompleted(state: CuriousState): Promise<boolean> {
    const last = state.history[state.history.length - 1];
    if (last?.phase !== "sync" || last.status !== "finished") {
      return false;
    }

    const status = analyzeRoadmap(
      await readFile(this.config.specPath, "utf8"),
    );
    if (!status.complete) {
      return false;
    }

    console.log(
      `[curious] roadmap complete — all ${status.totalTasks} tasks checked off`,
    );
    return true;
  }

  private printContinueHint(completedLimit: number | undefined): void {
    if (completedLimit === 1) {
      console.log("");
      console.log("[curious] To keep going through the roadmap:");
      console.log("  curious run              # until all roadmap tasks are done");
      console.log("  curious run --cycles 5   # five more full cycles");
      console.log("  curious run --cycle      # one more full cycle");
    }
  }

  private printNextStepHint(state: CuriousState): void {
    const last = state.history[state.history.length - 1];
    if (!last) return;

    if (last.status !== "finished") {
      console.log(
        `[curious] phase "${last.phase}" did not finish successfully — fix and re-run`,
      );
      return;
    }

    console.log(
      `[curious] phase "${last.phase}" done — next phase is "${state.phase}" (cycle ${state.cycle})`,
    );

    if (state.phase === "review") {
      console.log("[curious] tip: npm run review  OR  npm run run:cycle");
    } else if (state.phase === "sync") {
      console.log("[curious] tip: npm run sync  OR  npm run run:cycle");
    } else if (state.phase === "overseer") {
      console.log("[curious] tip: curious run  OR  npm run run");
    } else if (state.phase === "develop") {
      console.log("[curious] tip: npm run develop  OR  npm run run:cycle");
    }
  }
}

export async function printStatus(config: ResolvedConfig): Promise<void> {
  const state = await loadState(config.projectRoot);
  console.log(
    JSON.stringify(
      {
        config: {
          projectRoot: config.projectRoot,
          cwd: config.cwd,
          specPath: config.specPath,
          runtime: config.runtime,
        },
        state,
      },
      null,
      2,
    ),
  );
}

export async function resetState(config: ResolvedConfig): Promise<void> {
  await saveState(config.projectRoot, initialState());
  console.log("[curious] state reset");
}

File: python/src/curious/orchestrator.py
from __future__ import annotations

import signal
import time
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

from curious.config import ResolvedConfig, print_config_summary
from curious.workspace import prepare_agent_workspace
from curious.harness import run_harness
from curious.harness.best_of_n import develop_best_of_n
from curious.scanner_rules import (
    check_rules,
    format_violations_for_prompt,
    load_rules,
    save_rules,
    synthesize_rules,
)
from curious.spec_history import (
    OverseerLabel,
    append_overseer_label,
    count_diff_lines,
    overseer_diff_for_cycle,
    snapshot_agents,
    snapshot_spec,
)
from curious.spec_sections import extract_spec_section
from curious.review_verdict import parse_review_verdict
from curious.verifier.model import load_verifier, log_disagreement
from curious.orchestrator_cycles import (
    should_abort_cycles_mode_on_phase_error,
    should_stop_after_requested_cycles,
    should_stop_at_config_max_cycles,
)
from curious.orchestrator_roadmap import (
    should_skip_until_done_loop,
    should_stop_until_done_after_phase,
)
from curious.overseer import overseer_trigger_reason, should_run_overseer
from curious.project import load_agents_document, relative_to_root
from curious.prompts import build_prompt
from curious.spec_roadmap import analyze_roadmap
from curious.state import initial_state, load_state, next_phase, save_state
from curious.types import CycleRecord, CuriousState, Phase

REMAINING_TASK_PREVIEW = 8


@dataclass
class OrchestratorOptions:
    verbose: bool = False
    once: bool = False
    cycles: int | None = None
    until_done: bool = True


class Orchestrator:
    def __init__(self, config: ResolvedConfig, opts: OrchestratorOptions | None = None):
        self.config = config
        self.opts = opts or OrchestratorOptions()
        self.stopping = False
        self.last_summary: str | None = None

    def request_stop(self) -> None:
        self.stopping = True

    def run(self) -> None:
        prepare_agent_workspace(self.config)
        print_config_summary(self.config)
        state = load_state(self.config.project_root)
        state.running = True
        save_state(self.config.project_root, state)

        def shutdown(_signum=None, _frame=None) -> None:
            self.request_stop()
            nonlocal state
            state.running = False
            save_state(self.config.project_root, state)

        signal.signal(signal.SIGINT, shutdown)
        signal.signal(signal.SIGTERM, shutdown)

        try:
            cycle_at_start = state.cycle
            cycles_limit = self.opts.cycles
            skip_loop = False
            spec_path = Path(self.config.spec_path)

            if self.opts.once:
                print("[curious] mode: single phase (--once)")
            elif self.opts.until_done:
                initial = analyze_roadmap(spec_path.read_text(encoding="utf-8"))
                self._log_roadmap_status(initial)
                if should_skip_until_done_loop(initial):
                    print("[curious] roadmap already complete — nothing to do")
                    skip_loop = True
                else:
                    print(
                        "[curious] mode: until done (stops when all ## Roadmap tasks are checked)"
                    )
            elif cycles_limit is not None:
                print(
                    f"[curious] mode: {cycles_limit} full cycle(s), then stop"
                )
            else:
                print("[curious] mode: continuous (Ctrl+C to stop)")

            if not skip_loop:
                while not self.stopping:
                    state = self._run_phase(state)
                    if self.stopping:
                        break

                    if self.opts.once:
                        self._print_next_step_hint(state)
                        break

                    if should_stop_after_requested_cycles(
                        state, cycle_at_start, cycles_limit
                    ):
                        print(
                            f"[curious] finished {state.cycle - cycle_at_start} cycle(s)"
                        )
                        break

                    if self.opts.until_done and self._should_stop_until_done(state):
                        break

                    if should_abort_cycles_mode_on_phase_error(
                        cycles_limit, state.last_error
                    ):
                        print("[curious] cycle aborted due to phase error")
                        break

                    if should_stop_at_config_max_cycles(
                        state.cycle, self.config.max_cycles
                    ):
                        print(
                            f"[curious] reached maxCycles={self.config.max_cycles}"
                        )
                        break

                    if state.phase == "develop" and state.cycle > cycle_at_start:
                        print(
                            f"[curious] next: cycle {state.cycle} · develop"
                        )
                        if self.config.cycle_delay_ms > 0:
                            time.sleep(self.config.cycle_delay_ms / 1000)
        finally:
            state = load_state(self.config.project_root)
            state.running = False
            save_state(self.config.project_root, state)

    def _run_phase(self, state: CuriousState) -> CuriousState:
        phase = state.phase
        spec_body = Path(self.config.spec_path).read_text(encoding="utf-8")
        agents = load_agents_document(
            Path(self.config.project_root), Path(self.config.cwd)
        )

        if phase == "develop" and not agents:
            print(
                "[curious] warning: no AGENTS.md — developer lacks style guidelines"
            )
        elif agents and phase in ("develop", "review", "overseer"):
            print(f"[curious] including AGENTS.md ({agents.rel_path})")

        if phase == "overseer":
            print(
                f"[curious] overseer triggered: {overseer_trigger_reason(state, self.config)}"
            )

        root = Path(self.config.project_root)
        spec_sha = snapshot_spec(root, state.cycle, spec_body, phase)
        agents_sha = None
        if agents and agents.content:
            agents_sha = snapshot_agents(root, state.cycle, phase, agents.content)

        prompt = build_prompt(
            phase=phase,
            spec_path=self.config.spec_path,
            spec_rel_path=relative_to_root(
                Path(self.config.project_root), Path(self.config.spec_path)
            ),
            spec_body=spec_body,
            cycle=state.cycle,
            cwd=self.config.cwd,
            project_root=self.config.project_root,
            config=self.config,
            agents=agents,
            last_summary=self.last_summary,
            history=state.history,
        )

        if phase == "review":
            violations = check_rules(Path(self.config.cwd), load_rules(root))
            block = format_violations_for_prompt(violations)
            if block:
                prompt = prompt + "\n\n" + block

        llm = self.config.llm_for_phase(phase)
        spec_section = extract_spec_section(spec_body, "## Roadmap") or spec_body[:4000]
        agents_section = (agents.content[:4000] if agents else "")

        print(f"\n[curious] ── cycle {state.cycle} · {phase} ──")

        started_at = _now()
        run_id = ""
        status: str = "error"
        summary: str | None = None
        result = None

        try:
            if (
                phase == "develop"
                and self.config.harness.best_of_n.enabled
                and self.config.verifier.enabled
            ):
                verifier = load_verifier(
                    self.config.verifier.model_path,
                    self.config.verifier.base_model,
                )
                result = develop_best_of_n(
                    prompt,
                    Path(self.config.cwd),
                    llm,
                    self.config.harness,
                    verifier,
                    cycle=state.cycle,
                    spec_section=spec_section,
                    agents_section=agents_section,
                    verbose=self.opts.verbose,
                    bon=self.config.harness.best_of_n,
                )
            else:
                result = run_harness(
                    prompt,
                    Path(self.config.cwd),
                    llm,
                    self.config.harness,
                    verbose=self.opts.verbose,
                )
            run_id = result.run_id
            print(f"[curious] run {run_id}")
            status = result.status
            summary = result.summary
            self.last_summary = summary

            if result.status == "error":
                state.last_error = result.error or "harness error"
                print(f"[curious] run error: {state.last_error}")
            elif result.status == "finished":
                state.last_error = None
                preview = (summary or "")[:400]
                if len(summary or "") > 400:
                    preview += "…"
                print(f"[curious] result: {preview}")
        except Exception as exc:
            state.last_error = str(exc)
            print(f"[curious] phase error: {exc}")

        overseer_intervened = False
        if phase == "overseer" and status == "finished":
            after_spec = Path(self.config.spec_path).read_text(encoding="utf-8")
            diff = overseer_diff_for_cycle(root, state.cycle)
            if diff:
                added, removed = count_diff_lines(diff)
                sha_after = snapshot_spec(root, state.cycle, after_spec, phase)
                append_overseer_label(
                    root,
                    OverseerLabel(
                        cycle=state.cycle,
                        sha_before=spec_sha,
                        sha_after=sha_after,
                        lines_added=added,
                        lines_removed=removed,
                        summary=(summary or "")[:500],
                    ),
                )
                new_rules = synthesize_rules(
                    diff, spec_sha, state.cycle, self.config.llm
                )
                if new_rules:
                    existing = load_rules(root)
                    save_rules(root, existing + new_rules)
                overseer_intervened = True

        if (
            phase == "review"
            and status == "finished"
            and self.config.verifier.enabled
            and summary
        ):
            try:
                from curious.harvest.verifier import _git_diff

                verifier = load_verifier(
                    self.config.verifier.model_path,
                    self.config.verifier.base_model,
                )
                diff = _git_diff(self.config.project_root, self.config.cwd)
                scores = verifier.score(
                    diff=diff,
                    spec_section=spec_section,
                    agents_section=agents_section,
                )
                verdict = parse_review_verdict(summary)
                reviewer_pass = bool(verdict and verdict.overall == "PASS")
                log_disagreement(
                    root,
                    self.config.verifier.disagreement_log,
                    cycle=state.cycle,
                    verifier_scores=scores,
                    reviewer_pass=reviewer_pass,
                    diff_excerpt=diff,
                )
            except Exception as exc:
                print(f"[curious] verifier check skipped: {exc}")

        record = CycleRecord(
            cycle=state.cycle,
            phase=phase,
            run_id=run_id or "unknown",
            status=status,  # type: ignore[arg-type]
            started_at=started_at,
            finished_at=_now(),
            summary=summary,
            trajectory=result.trajectory or [] if result else [],
            spec_snapshot_sha=spec_sha,
            agents_snapshot_sha=agents_sha,
            overseer_intervened=overseer_intervened,
        )
        state.history.append(record)
        if len(state.history) > 100:
            state.history = state.history[-100:]
        state.last_run_id = run_id or state.last_run_id

        if status == "finished":
            if phase == "sync":
                state.cycle += 1
                state.phase = (
                    "overseer"
                    if should_run_overseer(state, self.config)
                    else next_phase("sync")
                )
            else:
                state.phase = next_phase(phase)
        else:
            print(
                f'[curious] staying on phase "{phase}" until a run finishes successfully'
            )

        save_state(self.config.project_root, state)
        return state

    def _should_stop_until_done(self, state: CuriousState) -> bool:
        spec_body = Path(self.config.spec_path).read_text(encoding="utf-8")
        if not should_stop_until_done_after_phase(state, spec_body):
            return False
        status = analyze_roadmap(spec_body)
        print(
            f"[curious] roadmap complete — all {status.total_tasks} tasks checked off"
        )
        return True

    def _log_roadmap_status(self, status) -> None:
        if status.total_tasks == 0:
            print("[curious] roadmap: no T*/M* tasks found")
            return
        remaining = status.unchecked_task_ids
        if len(remaining) > REMAINING_TASK_PREVIEW:
            head = ", ".join(remaining[:REMAINING_TASK_PREVIEW])
            rem = f", … (+{len(remaining) - REMAINING_TASK_PREVIEW} more)"
        else:
            head = ", ".join(remaining)
            rem = ""
        print(
            f"[curious] roadmap: {status.checked_tasks}/{status.total_tasks} done"
            + (f" (remaining: {head}{rem})" if head else "")
        )

    def _print_next_step_hint(self, state: CuriousState) -> None:
        if not state.history:
            return
        last = state.history[-1]
        if last.status != "finished":
            print(
                f'[curious] phase "{last.phase}" did not finish — fix and re-run'
            )
            return
        print(
            f'[curious] phase "{last.phase}" done — next: "{state.phase}" (cycle {state.cycle})'
        )


def print_status(config: ResolvedConfig) -> None:
    import json

    state = load_state(config.project_root)
    print(
        json.dumps(
            {
                "config": {
                    "projectRoot": config.project_root,
                    "cwd": config.cwd,
                    "specPath": config.spec_path,
                    "model": config.llm.model,
                },
                "state": state.to_json(),
            },
            indent=2,
        )
    )


def reset_state(config: ResolvedConfig) -> None:
    save_state(config.project_root, initial_state())
    print("[curious] state reset")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")
